### ISSUE TITLE
The status of the vacuum cleaner does not change when you return to the base.

### ISSUE THREAD

**Issue by earlywaff** [homebridge-xiaomi-roborock-vacuum/homebridge-xiaomi-roborock-vacuum#146]
When the vacuum cleaner finishes cleaning, it still writes in the Home Kit that the vacuum cleaner is on cleaning, restarting the application does not help. I press the vacuum cleaner switch several times after which the bridge crashes, press it several more times, restart the Home application and the robot-based vacuum cleaner.

Roborock S5

**Comment by afharo** [comment]
> I press the vacuum cleaner switch several times after which the bridge crashes

Does the bridge crash after you switch the cleaner ON? 😮

**Comment by afharo** [comment]
@earlywaff is this still valid?

**Comment by earlywaff** [comment]
I am sitting on version 0.7.2.  I can test on the new version.  but so far everything is working fine for me!)

**Comment by afharo** [comment]
Ok! If you are happy so far. Shall we close this issue?

**Comment by Bladestar2105** [comment]
Same issue here. After cleaning is finished the robot returns to the base but the status of the switch remains active!

**Comment by afharo** [comment]
I'm starting to think this may be related to `homebridge` itself and the underlying communication layer with the Home app. Reading through their [CHANGELOG](https://github.com/homebridge/homebridge/blob/master/CHANGELOG.md) they claim there was an issue with the encryption layer, making some updates to fail to be sent to Homekit.

I, myself, also suffered it for a while and thought it was due to a faulty connection when the module lost connection to the robot and kept failing to communicate ever since. But, since the last restart, it hasn't happened (59 days ago).

This issue happens quite randomly in this module to try and tackle it. If anyone finds a repeatable way to replicate the issue consistently so we can test it and find a proper way to auto-recovery itself, I'm afraid it's super hard to try to tackle it this blindly.

**Comment by bedrin** [comment]
Just reproduced this bug again. Robot has finished cleaning but fan is still shown as active

`cleaning` state is still reported as `true`
```
DEB getState | roborock.vacuum.e2 | State {"batteryLevel":100,"charging":false,"cleaning":true,"fanSpeed":68}
```

Here's the current state of robot:
```
/homebridge/node_modules/.bin # ./miio protocol call 192.168.1.34 get_status
 INFO  Attempting to call get_status on 192.168.1.34

 INFO  Device found, making call

 INFO  Got result:
[
  {
    "msg_ver": 8,
    "msg_seq": 111,
    "state": 8,
    "battery": 64,
    "clean_mode": 0,
    "fan_power": 68,
    "error_code": 0,
    "map_present": 1,
    "in_cleaning": 0,
    "dnd_enabled": 0,
    "begin_time": 1589280188,
    "clean_time": 3237,
    "clean_area": 48195000,
    "clean_trigger": 2,
    "back_trigger": 6,
    "completed": 1,
    "clean_strategy": 0
  }
]
```

**Comment by bedrin** [comment]
I'm new to both miIO and homebrdige but I think that robot reports status as `in_cleaning` while plugin is looking at `cleaning` which is only updated through the Apple Home but not when robot completes the cleaning.

**Comment by afharo** [comment]
The `cleaning` flag is set here: https://github.com/homebridge-xiaomi-roborock-vacuum/miio/blob/master/lib/devices/vacuum.js#L124-L168

Maybe we need to change https://github.com/homebridge-xiaomi-roborock-vacuum/miio/blob/master/lib/devices/vacuum.js#L105

To be like 
```
this.defineProperty('in_cleaning', value => {
  this.updateCleaning(value === 1);
  return value;
});
```
Or we can stop relying on the "cleaning" property in the state and read the 'in_cleaning' value instead.

**Comment by bedrin** [comment]
Looks like `propertyUpdated()` isn't called at all when robot finishes the cleaning. Hence `updateCharging()` isn't called either.

**Comment by afharo** [comment]
it would be great if we get to the bottom of this... But since it depends on connection drops, it's hard to replicate consistently to ease the tests :/

**Comment by bedrin** [comment]
It literally has happened in the middle of the night. I've added logging here:
https://github.com/homebridge-xiaomi-roborock-vacuum/miio/blob/master/lib/devices/vacuum.js#L53

And it stopped logging at around 4AM. No other log messages around it.

**Comment by bedrin** [comment]
Probably there were errors during polling. I am not 100% sure if logging works in this code:
https://github.com/tinkerhub/abstract-things/blob/master/polling.js#L55
```
this.debug('Could not poll:', ex);
```

`debug` implementation looks a bit weird to me. I've replaced it with simple `console.log` - let us see  what I get tomorrow if it crashes again over night.

I wonder if it could be some kind of power saving mode which kicks in

**Comment by afharo** [comment]
Thank you for taking ownership of that!
Fingers crossed for you to catch it! 🤞

**Comment by bedrin** [comment]
And it happened again at exactly the same time - around 3:30 AM my time
3 consequitive "Could not poll Error: Call to device timed out" errors

Afther that it has stopped polling the robot

**Comment by bedrin** [comment]
I think it's some kind of periodic job scheduled in robot which makes it inresponsive.
We can either increase the maximum number of failures for it or implement something else to workaround it. Not sure yet. Thoughts?

**Comment by afharo** [comment]
I'd go for not giving up on the re-connections maybe simply retrying with a backoff mechanism...

Or, even better, if the connection retry gives up, stop claiming it's connected so it will try to reconnect the next time a command is triggered (in the case of this module, every 30s): https://github.com/homebridge-xiaomi-roborock-vacuum/homebridge-xiaomi-roborock-vacuum/blob/60097be90c5be1adb1531a84e43fa5dc9a5ebf52/index.js#L632-L635

At the moment, I believe we don't have any other way to check if it's connected apart from getting the socket: https://github.com/homebridge-xiaomi-roborock-vacuum/homebridge-xiaomi-roborock-vacuum/blob/60097be90c5be1adb1531a84e43fa5dc9a5ebf52/index.js#L680

Maybe there's any additional option to validate it's connected?

**Comment by bedrin** [comment]
> Maybe there's any additional option to validate it's connected?

We can call `get_status` for example. Not sure if it's implemented on all devices though.

I think `this.device.state()` doesn't actually make a request to device and instead returns an internal state from miio library.

**Comment by afharo** [comment]
Maybe we can call `this.device.poll()`? 
https://github.com/homebridge-xiaomi-roborock-vacuum/miio/blob/d01af70ea0a32c3570d311db855f23ff739c2d23/lib/device.js#L135

It internally calls `loadProperties`, calling `get_status` and updating all the props for us :)
https://github.com/homebridge-xiaomi-roborock-vacuum/miio/blob/master/lib/devices/vacuum.js#L287-L308

This way, if we have other models that have other methods (#150), we only need to rename the internal methods, but not the homebridge module

**Comment by bedrin** [comment]
I like it. So basically we'll have two pollings simultaneously, right? 
Our own and the one implemented inside miio/abstract-things libraries.

**Comment by afharo** [comment]
Yup! We can add `await this.device.poll()` right before this line:
https://github.com/homebridge-xiaomi-roborock-vacuum/homebridge-xiaomi-roborock-vacuum/blob/60097be90c5be1adb1531a84e43fa5dc9a5ebf52/index.js#L703

And it will be called every 30s :)

**Comment by bedrin** [comment]
I'm gonna give it a try. Also today found another issue - when I remove the dust bin I'm getting error 9.
However when I install it back homebridge-xiaomi-roborock-vacuum still reports it as 9

Looks like a bug - will raise it separatelly when I have a chance.

**Comment by afharo** [comment]
I think it might be related to this error: https://github.com/homebridge-xiaomi-roborock-vacuum/homebridge-xiaomi-roborock-vacuum/issues/193

**Comment by afharo** [comment]
I tried fixing it #173 but apparently it didn't work as expected... :/

**Comment by afharo** [comment]
maybe the `poll()` => `loadProperties` will enforce the property `error` to refresh as well 🤷

**Comment by bedrin** [comment]
Looks like `await this.device.poll()` has fixed this issue - I'm gonna raise a PR for it.
However https://github.com/homebridge-xiaomi-roborock-vacuum/homebridge-xiaomi-roborock-vacuum/issues/193 is still there.

I can see that loadProperties does return `error_code` = 0 however logs in `homebridge-xiaomi-roborock-vacuum` (WAR changedError ...) continue showing an error even when it's resolved

**Comment by afharo** [comment]
Brilliant news!!

Re the repeated errors, we can discuss it later :)

**Comment by bedrin** [comment]
> Brilliant news!!
https://github.com/homebridge-xiaomi-roborock-vacuum/homebridge-xiaomi-roborock-vacuum/pull/194
 
> Re the repeated errors, we can discuss it later :)

Yeah, I think it only affects logs in any case.

Might be caused by this guy at the end of `getState` method:
```
safeCall(state.error, (error) => this.changedError(error));
```

But indeed it's a separate discussion :)

**Comment by adamrzanek** [comment]
Same problem here. Additionally most of the time roborock doesn’t clean when triggered by time automation in HomeKit.

**Comment by afharo** [comment]
@adamrzanek can you try if the beta version `homebridge-xiaomi-roborock-vacuum@next` fixes the issue for you? Thank you!

**Comment by afharo** [comment]
It should be fixed in the latest `0.9.0` release. Please, feel free to reopen if still happening
